### DISCUSSION TITLE
gcloud backend

### DISCUSSION THREAD

**Discussion by glasser** [jdx/mise#3858] [Ideas]
This is somewhat discussed in #3709 as a broader concept of plugin backends, but I figured it would be nice to break this out as a potential standalone backend, even just to be potentially rejected.

Concrete issue: in order to use kubectl to interact with Google Cloud Kubernetes clusters, you need to have gke-gcloud-auth-plugin installed, so I would like my company's project repositories to install it via mise along with the rest of our dev environment. This tool is only officially distributed as a "component" of the gcloud CLI, which you install by running a `gcloud components install` command after installing the CLI.

https://github.com/jthegedus/asdf-gcloud exists and does allow you to provide a file listing components to install with gcloud but it seems more designed for tracking that list at a single global homedir level, not per repository. It would be great to have a more direct way to use mise to ensure that gke-gcloud-auth-plugin is available in a directory.

**Comment by glasser** [comment]
(it's possible this can be solved fully via hooks — haven't quite got it working but gonna try a bit harder.)

**Reply by glasser** [reply]
I don't feel like I've gotten this to be solved in a nice enough way via hooks.

**Reply by glasser** [reply]
The main thing I'm missing with hooks is a way to ensure `gcloud components install gke-gcloud-auth-plugin` gets called when the active tool version of `gcloud` changes, whether or not the version was already installed on disk (perhaps by some other repo that doesn't care about gke-gcloud-auth-plugin).  An `enter` hook kinda helps but not if you're jumping around between git commits that change the tool version without leaving the directory.  Some sort of "post-toolset-change" hook, or a way to do a tool-specific "post-add-to-path" hook like the tool-specific "postinstall" hook that exists would be nice.

**Comment by PaulBemowski-dematic** [comment]
I am struggling with the same issue.  One way to solve it (which is distasteful) is to create a static durable installation of google-cloud-sdk, install the auth plugin - and then set the PATH variable in subtrees where you want kubectl access - but that really defeats the usefulness of mise.

I wondered if you solved this problem in a more elegant way since last year?

**Reply by PaulBemowski-dematic** [reply]
I was able to solve it with this:

```
[tools]
gcloud = {version='568.0.0', postinstall="gcloud components install gke-gcloud-auth-plugin"}
```

**Reply by glasser** [reply]
Last I checked that only works if this particular line is what causes that particular version to be installed — if it gets installed somewhere else and then this gets evaluated, the script doesn't run.

**Reply by flou** [reply]
I had the same issue which I tried to solve by updating the post-install hook of the vfox-gcloud plugin, I sent a PR on the plugin repository https://github.com/mise-plugins/vfox-gcloud/pull/5
